Replace custom cmake code with ign-cmake2

scpeters opened the pull request:
# 🎉 New feature

Closes #181, borrowing from #747 and targeting the `sdf10` branch. It should be possible to backport these changes to `sdf9` if an appropriate find-module is provided for TinyXML.

## Summary

This replaces most of the custom cmake code in libsdformat with the functionality provided by `ignition-cmake2`. The root `CMakeLists.txt` is much shorter now and most of the `cmake` folder has been deleted. This is made possible by the `NO_IGNITION_PREFIX` and `REPLACE_IGNITION_INCLUDE_PATH` parameters added to `ign_configure_project` in https://github.com/ignitionrobotics/ign-cmake/pull/190 and https://github.com/ignitionrobotics/ign-cmake/pull/191. Some other details are listed below:

* Use `ign-cmake` find modules for Tinyxml2 and URDFDOM (added in https://github.com/ignitionrobotics/ign-cmake/pull/193), which provide cmake targets
* the `sdf_config.h.in` template is renamed to `config.hh.in` to match the ign-cmake convention, and an `sdf_config.h` static header is added to preserve API
* the autogenerated header is now `sdformat.hh` instead of `sdf/sdf.hh`, so a static header `sdf/sdf.hh` is added to preserve API
* the `SDFORMAT_*` visibility macros in `system_util.h` are now based on the `IGNITION_SDFORMAT_*` macros from the auto-generated `sdf/Export.hh`
* the `LEGACY_PROJECT_PREFIX ` option from https://github.com/ignitionrobotics/ign-cmake/pull/199 is used to ensure the legacy cmake variables from the cmake config file retain the `SDFormat_` prefix

~~This is currently a draft pull request since~~ There are some differences between the output of this branch and the currently installed versions:

* headers are installed to `{prefix}/include/ignition/sdformat10` rather than `{prefix}/include/sdformat-10.6`. EDIT: I think this is ok because the header folder changes with each minor release anyways. As long as the cmake config and pkgconfig files point to the correct place, it should be fine.
* the name / description in the `.pc` file have changed from `SDF` / `Robot Modeling Language (SDF)` to `Ignition sdformat` / `A set of sdformat classes for robot applications`
* the `SDFORMAT_RUN_VALGRIND_TESTS` cmake option is removed since `ign_build_tests` does not support valgrind

There ~~are~~ is also ~~two~~ one remaining custom cmake function in the `cmake` folder:

* `find_python_module`: this function is used to find the `psutil` python module that is used by an integration test.

## Test it

It should continue to work normally. Testing in a colcon workspace with other packages would be very helpful.

## Checklist
- [X] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [X] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [X] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

scpeters commented:
testing https://github.com/ignition-tooling/action-ignition-ci/pull/46

scpeters commented:
the ABI-checker needs https://github.com/ignition-tooling/release-tools/pull/589

scpeters commented:
@osrf-jenkins run tests please

scpeters commented:
the warnings in the Ubuntu build should be fixed by https://github.com/ignition-tooling/release-tools/pull/590

scpeters commented:
> * the `SDFormat_*` cmake variables are no longer provided by our cmake configuration file, which breaks our example CMakeLists.txt (see results of bionic GitHub Actions workflow)

the `SDFormat_LIBRARIES` variable was still in use in ignition-physics, so I'm updating it to use the proper cmake targets in https://github.com/ignitionrobotics/ign-physics/pull/314

koonpeng commented:
There are some additional changes not in #747, will this be automatically forward ported to other branches or should I start cherry-picking the changes.

scpeters commented:
> There are some additional changes not in #747, will this be automatically forward ported to other branches or should I start cherry-picking the changes.

It will be forward-ported. I believe the most relevant branch for https://github.com/ignitionrobotics/sdformat/pull/762 is `sdf12`

chapulina reviewed (commented):
Nice! Just some questions:

* Do we still need `cmake/SDFUtils.cmake` and `sdf_build_tests`?
* Have you tried to generate and upload docs to make sure it still works?
* Do we still need all the scripts in `tools`?

scpeters commented:
> Nice! Just some questions:
> 
> * Do we still need `cmake/SDFUtils.cmake` and `sdf_build_tests`?

from the PR description:

* "sdf_build_tests: this macro has a parameter `SDF_BUILD_TESTS_EXTRA_EXE_SRCS` that allows extra source files to be compiled with a test. This is used by several unit tests in `src/CMakeLists.txt` that use symbols from private header files that aren't publicly exported"

> * Have you tried to generate and upload docs to make sure it still works?

I have not tested this

> * Do we still need all the scripts in `tools`?

we didn't remove the codecheck scripts in https://github.com/ignitionrobotics/sdformat/pull/682, though I suppose we could consider doing that. We are definitely using the following two scripts:

* https://github.com/ignitionrobotics/sdformat/blob/sdf12/tools/get_mem_info.py
* https://github.com/ignitionrobotics/sdformat/blob/sdf12/tools/xmlschema.rb

adlarkin commented on examples/CMakeLists.txt:
```diff
@@ -4,11 +4,8 @@ find_package(sdformat10 REQUIRED)
 add_executable(simple simple.cc)
-target_link_libraries(simple ${SDFormat_LIBRARIES})
+target_link_libraries(simple sdformat10::sdformat10)
 
 add_executable(dom dom.cc)
-target_link_libraries(dom ${SDFormat_LIBRARIES})
+target_link_libraries(dom sdformat10::sdformat10)
```
What if we set the SDF version (in this case, 10) to a variable instead? That way, if these changes were to be ported across different versions, updating the version would be easier since we would only need to update the variable's value.

scpeters replied:
fixed in https://github.com/ignitionrobotics/sdformat/pull/780/commits/e1917735792647c8459c9a1c3cdcffec4ec3f5a3

adlarkin commented on include/sdf/sdf.hh:
```diff
@@ -0,0 +1,26 @@
+#ifndef SDF_SDF_HH_
+#define SDF_SDF_HH_
+
+// TODO(anyone): deprecate this header?
```
@chapulina mentioned the other day/week that according to the google c++ style guide, the name in a `TODO` should actually be the name of the person who created it, since this person knows the most about the `TODO`: https://google.github.io/styleguide/cppguide.html#TODO_Comments

I know that in the past, we've been making the name in a `TODO` the name of someone who should fix it, but we should change this to adhere more closely to the style guide - @scpeters, would you mind putting your name in the `TODO` since you're the one who created this `TODO` and knows the most about it?

scpeters replied:
I copied it from https://github.com/ignitionrobotics/sdformat/pull/747, but I'll just remove the comment since it's not strictly necessary

scpeters replied:
removed the comment in https://github.com/ignitionrobotics/sdformat/pull/780/commits/a13f9e5e65f8f941619eb9dd41324bad89981875

adlarkin commented on include/sdf/sdf_config.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef SDF_SDF_CONFIG_H_
+#define SDF_SDF_CONFIG_H_
+
+// TODO(anyone): deprecate this header?
```
Same comments about changing the name in the `TODO` note apply here

scpeters replied:
removed the comment in https://github.com/ignitionrobotics/sdformat/pull/780/commits/a13f9e5e65f8f941619eb9dd41324bad89981875

scpeters commented on CMakeLists.txt:
```diff
@@ -4,350 +4,127 @@ if(COMMAND CMAKE_POLICY)
-set(sdf_import_target_name ${PROJECT_EXPORT_NAME}::${sdf_target})
-set(sdf_target_output_filename "${sdf_target}-targets.cmake")
-
+message (STATUS "${PROJECT_NAME} version ${PROJECT_VERSION_FULL}")
```
good catch, this line can be removed. It's a duplicate of the following line, which does have the variable properly defined:

* https://github.com/ignitionrobotics/ign-cmake/blob/ign-cmake2/cmake/IgnConfigureProject.cmake#L112

scpeters replied:
removed in https://github.com/ignitionrobotics/sdformat/pull/780/commits/e1917735792647c8459c9a1c3cdcffec4ec3f5a3

scpeters commented on CMakeLists.txt:
```diff
@@ -4,350 +4,127 @@ if(COMMAND CMAKE_POLICY)
-  endif("${isSystemDir}" STREQUAL "-1")
-endif()
-
 set (BUILD_SDF ON CACHE INTERNAL "Build SDF" FORCE)
```
I left this in place for the principle of backwards compatibility, but I'm not sure if anyone uses it. I believe it was copied from `gazebo`, which still has a similar `BUILD_GAZEBO` cmake option:

* https://github.com/osrf/gazebo/blob/gazebo11_11.9.1/CMakeLists.txt#L132
* https://github.com/osrf/gazebo/blob/gazebo11_11.9.1/CMakeLists.txt#L381-L398

adlarkin replied:
Thanks for the explanation :+1: I think leaving it in doesn't hurt - @chapulina, do you have any strong thoughts/opinions about this?

j-rivero replied:
```suggestion
# BUILD_SDF is preserved for backwards compatibility but can be removed in main branch
set (BUILD_SDF ON CACHE INTERNAL "Build SDF" FORCE)
```

chapulina replied:
Just saw this, leaving it there for backwards compatibility and removing on `main` sounds good to me :+1:

scpeters commented on CMakeLists.txt:
```diff
@@ -4,350 +4,127 @@ if(COMMAND CMAKE_POLICY)
+  if (NOT RUBY)
+      ign_build_error ("Ruby version 1.9 is needed to build xml schemas")
+  else()
+      message(STATUS "Found ruby executable: ${RUBY}")
```
fixed in https://github.com/ignitionrobotics/sdformat/pull/780/commits/e1917735792647c8459c9a1c3cdcffec4ec3f5a3

scpeters commented on CMakeLists.txt:
```diff
@@ -4,350 +4,127 @@ if(COMMAND CMAKE_POLICY)
+  # Find psutil python package for memory tests
+  include (${sdf_cmake_dir}/SDFUtils.cmake)
+  find_python_module(psutil)
+  if(NOT PY_PSUTIL)
```
fixed in https://github.com/ignitionrobotics/sdformat/pull/780/commits/e1917735792647c8459c9a1c3cdcffec4ec3f5a3

scpeters commented on CMakeLists.txt:
```diff
@@ -4,350 +4,127 @@ if(COMMAND CMAKE_POLICY)
-project (sdformat10)
+project (sdformat10 VERSION 10.6.0)
 
 # The protocol version has nothing to do with the package version set below.
```
fixed in https://github.com/ignitionrobotics/sdformat/pull/780/commits/e1917735792647c8459c9a1c3cdcffec4ec3f5a3

scpeters commented:
it looks like none of the unit tests are running. maybe it is a problem to mix `ign_build_tests` and `sdf_build_tests`

scpeters commented:
> it looks like none of the unit tests are running. maybe it is a problem to mix `ign_build_tests` and `sdf_build_tests`

should be fixed by https://github.com/ignitionrobotics/sdformat/pull/780/commits/a1d7bc530c36cd6fc56e9b46ab166b13606105e2

scpeters commented:
> > Nice! Just some questions:
> > 
> > * Do we still need `cmake/SDFUtils.cmake` and `sdf_build_tests`?
> 
> from the PR description:
> 
> * "sdf_build_tests: this macro has a parameter `SDF_BUILD_TESTS_EXTRA_EXE_SRCS` that allows extra source files to be compiled with a test. This is used by several unit tests in `src/CMakeLists.txt` that use symbols from private header files that aren't publicly exported"

well well well...I was seeing some test failures in GitHub Actions, and debugging led me to discover that the tests build just fine with `ign_build_tests` without needing the `EXTRA_EXE_SRCS` hack. I'm not sure what has changed but I'll take it. I've removed `sdf_build_tests` and `tools/check_test_ran.py` in https://github.com/ignitionrobotics/sdformat/pull/780/commits/910dddf161b7396cd20274b09387d75ef7d72670

> > * Do we still need all the scripts in `tools`?
> 
> we didn't remove the codecheck scripts in #682, though I suppose we could consider doing that. We are definitely using the following two scripts:
> 
> * https://github.com/ignitionrobotics/sdformat/blob/sdf12/tools/get_mem_info.py
> * https://github.com/ignitionrobotics/sdformat/blob/sdf12/tools/xmlschema.rb

I removed the cppcheck and cpplint scripts in https://github.com/ignitionrobotics/sdformat/pull/780/commits/62bc3440ae79e0f8813333ef9af822f4e4d858b9

koonpeng commented:
> well well well...I was seeing some test failures in GitHub Actions, and debugging led me to discover that the tests build just fine with ign_build_tests without needing the EXTRA_EXE_SRCS hack. I'm not sure what has changed but I'll take it. I've removed sdf_build_tests and tools/check_test_ran.py in 910dddf

That explains why it works on my branch (also the missing unittest seems to be due to some stale cmake cache so it still works on my pc). Upon further investigation, it works because ign-cmake automatically links the core library into the tests, I think a possible edge case would be where the test targets are not public apis (not exported), which would cause them to fail. But there doesn't seem to be any such cases so its all good.

azeey commented:
> well well well...I was seeing some test failures in GitHub Actions, and debugging led me to discover that the tests build just fine with ign_build_tests without needing the EXTRA_EXE_SRCS hack. I'm not sure what has changed but I'll take it. I've removed sdf_build_tests and tools/check_test_ran.py in 910dddf

The old CMake build enables `-fvisibility=hidden` on all translation units, thus, unless a symbol is explicitly set to be visible, it is hidden by default. This is why we needed the `EXTRA_EXE_SRCS` because the symbols in those files were not exported. With the new approach, all symbols are visible unless they are hidden explicitly (see https://github.com/ignitionrobotics/ign-cmake/issues/166). The downside of this is that it expands our ABI surface area, and even if we went through the code and added `SDFORMAT_HIDDEN` everywhere, it is easy to forget to do that every time we make changes and so we'd be prone to accidentally exporting symbols that we didn't mean to.

In https://github.com/ignitionrobotics/ign-cmake/issues/166, there's a proposal to change the behavior for ign-cmake3. What if we make it configurable like we did for `NO_IGNITION_PREFIX`?

scpeters commented:
> The old CMake build enables `-fvisibility=hidden` on all translation units, thus, unless a symbol is explicitly set to be visible, it is hidden by default. This is why we needed the `EXTRA_EXE_SRCS` because the symbols in those files were not exported. With the new approach, all symbols are visible unless they are hidden explicitly (see [ignitionrobotics/ign-cmake#166](https://github.com/ignitionrobotics/ign-cmake/issues/166)). The downside of this is that it expands our ABI surface area, and even if we went through the code and added `SDFORMAT_HIDDEN` everywhere, it is easy to forget to do that every time we make changes and so we'd be prone to accidentally exporting symbols that we didn't mean to.
> 
> In [ignitionrobotics/ign-cmake#166](https://github.com/ignitionrobotics/ign-cmake/issues/166), there's a proposal to change the behavior for ign-cmake3. What if we make it configurable like we did for `NO_IGNITION_PREFIX`?

great idea! I've added a cmake option in https://github.com/ignitionrobotics/ign-cmake/pull/196 that should provide the behavior desired by https://github.com/ignitionrobotics/ign-cmake/issues/166 and have used it in https://github.com/ignitionrobotics/sdformat/pull/780/commits/3a37dc8ad62dcefa3f36a696d97e140d6eaedabe. It turns out we don't need the `EXTRA_EXE_SRCS` hack if we use `target_sources` (which I just learned about).

scpeters commented:
> great idea! I've added a cmake option in [ignitionrobotics/ign-cmake#196](https://github.com/ignitionrobotics/ign-cmake/pull/196) that should provide the behavior desired by [ignitionrobotics/ign-cmake#166](https://github.com/ignitionrobotics/ign-cmake/issues/166) and have used it in [3a37dc8](https://github.com/ignitionrobotics/sdformat/commit/3a37dc8ad62dcefa3f36a696d97e140d6eaedabe). It turns out we don't need the `EXTRA_EXE_SRCS` hack if we use `target_sources` (which I just learned about).

I've merged https://github.com/ignitionrobotics/ign-cmake/pull/196 and made another prerelease and restarted Ubuntu CI

scpeters commented:
> > great idea! I've added a cmake option in [ignitionrobotics/ign-cmake#196](https://github.com/ignitionrobotics/ign-cmake/pull/196) that should provide the behavior desired by [ignitionrobotics/ign-cmake#166](https://github.com/ignitionrobotics/ign-cmake/issues/166) and have used it in [3a37dc8](https://github.com/ignitionrobotics/sdformat/commit/3a37dc8ad62dcefa3f36a696d97e140d6eaedabe). It turns out we don't need the `EXTRA_EXE_SRCS` hack if we use `target_sources` (which I just learned about).
> 
> I've merged [ignitionrobotics/ign-cmake#196](https://github.com/ignitionrobotics/ign-cmake/pull/196) and made another prerelease and restarted Ubuntu CI

CI is 🟢 again

scpeters commented:
@osrf-jenkins run tests please

(I want to rerun the homebrew build just to be sure)

azeey commented:
I would consider the removal of the `SDFormat_*` variables from the export list a breaking change since downstream packages that used those variables would now fail to build. Would it be possible to add those back, perhaps with another change in ign-cmake? It would be nice if ign-cmake had an API to export arbitrary additional variables.

scpeters commented:
> I would consider the removal of the `SDFormat_*` variables from the export list a breaking change since downstream packages that used those variables would now fail to build. Would it be possible to add those back, perhaps with another change in ign-cmake? It would be nice if ign-cmake had an API to export arbitrary additional variables.

I've opened https://github.com/ignitionrobotics/ign-cmake/pull/199 and committed https://github.com/ignitionrobotics/sdformat/pull/780/commits/e987976dfa28ef532dd6a2860232c5ce5cb2e678 in this branch to preserve the `SDFormat_*` variables

scpeters commented:
> I've opened [ignitionrobotics/ign-cmake#199](https://github.com/ignitionrobotics/ign-cmake/pull/199) and committed [e987976](https://github.com/ignitionrobotics/sdformat/commit/e987976dfa28ef532dd6a2860232c5ce5cb2e678) in this branch to preserve the `SDFormat_*` variables

https://github.com/ignitionrobotics/ign-cmake/pull/199 is merged and I'm preparing for another ign-cmake2 prerelease: https://github.com/ignitionrobotics/ign-cmake/pull/200

scpeters commented:
> > I've opened [ignitionrobotics/ign-cmake#199](https://github.com/ignitionrobotics/ign-cmake/pull/199) and committed [e987976](https://github.com/ignitionrobotics/sdformat/commit/e987976dfa28ef532dd6a2860232c5ce5cb2e678) in this branch to preserve the `SDFormat_*` variables
> 
> [ignitionrobotics/ign-cmake#199](https://github.com/ignitionrobotics/ign-cmake/pull/199) is merged and I'm preparing for another ign-cmake2 prerelease: [ignitionrobotics/ign-cmake#200](https://github.com/ignitionrobotics/ign-cmake/pull/200)

I've made the prerelease, waited for the debs to build, and have restarted the Ubuntu actions jobs. I am still waiting for the homebrew-simulation bottle to build

scpeters commented:
testing colcon windows builds with this branch:

* [![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ign_physics-ign-3-win&build=35)](https://build.osrfoundation.org/view/ign-dome/job/ign_physics-ign-3-win/35/) https://build.osrfoundation.org/view/ign-dome/job/ign_physics-ign-3-win/35/
* [![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ign_sensors-ign-4-win&build=26)](https://build.osrfoundation.org/view/ign-dome/job/ign_sensors-ign-4-win/26/) https://build.osrfoundation.org/view/ign-dome/job/ign_sensors-ign-4-win/26/

chapulina commented:
It would be interesting to run some CI builds for downstream libraries on each platform using this branch to make sure we're not breaking any expectations, is that relatively trivial to do?

j-rivero commented on CMakeLists.txt:
```diff
@@ -4,350 +4,109 @@ if(COMMAND CMAKE_POLICY)
+  ################################################
+  # Find the Python interpreter for running the
+  # check_test_ran.py script
+  find_package(PythonInterp 3 QUIET)
```
We need to maintain this deprecated CMake command to support BIonic 3.10.2 version of CMake but should be migrated if new versions don't need to support it.

scpeters replied:
actually, we don't need this at all since ign-cmake is providing `check_test_ran.py`

I'll remove it

j-rivero replied:
perfect.

scpeters replied:
removed in https://github.com/ignitionrobotics/sdformat/commit/795e10192d0a03859fab5fbe5c719fa2abd932ba

j-rivero commented on CMakeLists.txt:
```diff
@@ -4,350 +4,109 @@ if(COMMAND CMAKE_POLICY)
-    ${CMAKE_BINARY_DIR}/upload_doc.sh @ONLY)
+  ################################################
+  # Find ruby executable to produce xml schemas
+  find_program(RUBY ruby)
```
Would be a bit more elegant to use the `find_package(ruby)` + `RUBY_EXECUTABLE`, could cover some corner cases. Not a big deal anyway, leave it up to you.

scpeters replied:
Actually, I think it might make things harder. I believe all we need is the ruby executable, but I think you can only `find_package(ruby)` if you have the development package with ruby headers installed. For our purposes, I think `find_program` gives us what we need and is easier to satisfy

j-rivero replied:
makes sense, thanks.

j-rivero reviewed (changes requested):
Some comments on the main `CMakeLists.txt` file. I continue with the rest. The work is pretty cool :)

scpeters commented:
> It would be interesting to run some CI builds for downstream libraries on each platform using this branch to make sure we're not breaking any expectations, is that relatively trivial to do?

I ran windows builds for ign-physics and ign-sensors against this branch since that was easy to do:

* [![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ign_physics-ign-3-win&build=35)](https://build.osrfoundation.org/view/ign-dome/job/ign_physics-ign-3-win/35/) https://build.osrfoundation.org/view/ign-dome/job/ign_physics-ign-3-win/35/
* [![Build Status](https://build.osrfoundation.org/buildStatus/icon?job=ign_sensors-ign-4-win&build=26)](https://build.osrfoundation.org/view/ign-dome/job/ign_sensors-ign-4-win/26/) https://build.osrfoundation.org/view/ign-dome/job/ign_sensors-ign-4-win/26/

it's not so easy to do it for other platforms but I'll try

scpeters commented on CMakeLists.txt:
```diff
@@ -4,350 +4,109 @@ if(COMMAND CMAKE_POLICY)
-    ${LIB_INSTALL_DIR}/pkgconfig COMPONENT pkgconfig)
+  #################################################
+  # Find ign command line utility:
+  ign_find_package(ignition-tools)
```
I can replace this line with `find_program(HAVE_IGN_TOOLS ign)`; that makes sense to me

this flag is checked in the `sdf` folders as the `ign sdf -d` command is used to generate SDFormat description files. I'll update the comments around that cmake code to clarify

scpeters replied:
updated in https://github.com/ignitionrobotics/sdformat/pull/780/commits/c1ae0a3b3ddf5ab922c89a19ff6dcf1da451e3e3

j-rivero reviewed (commented):
Minor comments added. Overall looks great, a wonderful work. I'm not able to make it fail. As a final layout compatibility test we could try generating a nightly from this branch to check that files are named and located where expected.

I really like the changes guys.

scpeters commented on CMakeLists.txt:
```diff
@@ -4,350 +4,109 @@ if(COMMAND CMAKE_POLICY)
+    REPLACE_IGNITION_INCLUDE_PATH sdf)
+
+  message (STATUS "\n\n====== Finding 3rd Party Packages ======")
+  message(STATUS "Building for arch: ${ARCH}")
```
removed in https://github.com/ignitionrobotics/sdformat/pull/780/commits/795e10192d0a03859fab5fbe5c719fa2abd932ba

scpeters commented:
> Minor comments added. Overall looks great, a wonderful work. I'm not able to make it fail. As a final layout compatibility test we could try generating a nightly from this branch to check that files are named and located where expected.
> 
> I really like the changes guys.

what if we make a prerelease from `sdf10` before a stable release?

scpeters commented:
since it looks like no more changes are needed from ign-cmake, I'll open a PR for a stable 2.10.0 release

scpeters commented:
> since it looks like no more changes are needed from ign-cmake, I'll open a PR for a stable 2.10.0 release

https://github.com/ignitionrobotics/ign-cmake/pull/201

scpeters commented on include/sdf/sdf.hh:
```diff
@@ -0,0 +1,23 @@
+
+#ifndef SDF_SDF_HH_
+#define SDF_SDF_HH_
+
```
https://github.com/ignitionrobotics/sdformat/pull/780/commits/6865f453b3bd777179e151c6d740171d1b8b63d7

scpeters commented on include/sdf/sdf_config.h:
```diff
@@ -0,0 +1,23 @@
+
+#ifndef SDF_SDF_CONFIG_H_
+#define SDF_SDF_CONFIG_H_
+
```
https://github.com/ignitionrobotics/sdformat/pull/780/commits/6865f453b3bd777179e151c6d740171d1b8b63d7

scpeters commented on cmake/SDFUtils.cmake:
```diff
@@ -1,231 +1,3 @@
-
 #################################################
 # Copied from catkin/cmake/empy.cmake
 function(find_python_module module)
```
I moved it to the root CMakeLists.txt in https://github.com/ignitionrobotics/sdformat/pull/780/commits/4a7e7e3650fa471c7cd8ad149ac29e8290619244

I had to add `include(IgnPython)` to ensure that it finds python3

scpeters commented:
preparing sdformat10-release debian metadata:

* https://github.com/ignition-release/sdformat10-release/pull/6

scpeters commented:
> > since it looks like no more changes are needed from ign-cmake, I'll open a PR for a stable 2.10.0 release
> 
> [ignitionrobotics/ign-cmake#201](https://github.com/ignitionrobotics/ign-cmake/pull/201)

the release is done and I just deleted the matching branch on `gzdev` that was enabling prerelease repositories, so it should build from stable packages now

scpeters commented:
@osrf-jenkins run tests please

chapulina commented on cmake/upload_doc.sh.in:
```diff
@@ -39,11 +39,11 @@ if [ ! -f "@CMAKE_BINARY_DIR@/doxygen/html/index.html" ]; then
 
 if [ "$ans" = "y" ] || [ "$ans" = "Y" ]; then
-  s3cmd sync @CMAKE_BINARY_DIR@/doxygen/html/* s3://osrf-distributions/sdformat/api/@SDF_VERSION_FULL@/ -v
+  s3cmd sync @CMAKE_BINARY_DIR@/doxygen/html/* s3://osrf-distributions/sdformat/api/@PROJECT_VERSION_FULL@/ -v
```
Just checking, did anyone try uploading docs from this branch?

scpeters replied:
no, sorry I haven't tried that yet

chapulina reviewed (approved):
I just took a really quick look and didn't see anything fishy. I think we can merge this and make a pre-release to see if anything breaks downstream.

The only thing to keep in mind is that we should make the last 10.X release in the coming days for Dome's EOL. So in the worst case scenario, if we identify issues after the pre-release, we may need to revert this PR soon.

koonpeng commented:
@ahcorde @scpeters how do i go about forward porting this to `main`? I wans't able to find any docs on this.

scpeters commented:
> @ahcorde @scpeters how do i go about forward porting this to `main`? I wans't able to find any docs on this.

we forward port by merging forward and resolving conflicts. I'm testing a forward port to sdf11

chapulina commented:
> how do i go about forward porting this to main? I wans't able to find any docs on this.

It's documented here: https://ignitionrobotics.org/docs/all/contributing#process

scpeters commented:
> > @ahcorde @scpeters how do i go about forward porting this to `main`? I wans't able to find any docs on this.
> 
> we forward port by merging forward and resolving conflicts. I'm testing a forward port to sdf11

merging #789 forward first since its test needs a few changes with SDFormat 1.8: https://github.com/ignitionrobotics/sdformat/pull/807

scpeters commented:
> > > @ahcorde @scpeters how do i go about forward porting this to `main`? I wans't able to find any docs on this.
> > 
> > 
> > we forward port by merging forward and resolving conflicts. I'm testing a forward port to sdf11
> 
> merging #789 forward first since its test needs a few changes with SDFormat 1.8: #807

merging the rest forward to `sdf11` in https://github.com/ignitionrobotics/sdformat/pull/808

osrf-triage commented:
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-03-25-fortress-edifice-citadel/1343/1

osrf-triage commented:
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-04-13-fortress-edifice/1367/1
